SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,615 @@
+#include "LoadBalancing/InterestManager.h"
+#include "Schema/ChangeInterest.h"
+#include "Utils/InterestFactory.h"
+
+// Try to grab improbable's phTree thingy
+
+#include <algorithm>
+#include <cmath>
+
+#include <immintrin.h>
+
+DECLARE_CYCLE_STAT(TEXT("InterestManagerCompute"), STAT_InterestManagerComputation, STATGROUP_SpatialNet);
+DECLARE_CYCLE_STAT(TEXT("InterestManagerComputeBroadphase"), STAT_InterestManagerComputationBroadphase, STATGROUP_SpatialNet);
+DECLARE_CYCLE_STAT(TEXT("InterestManagerComputeBox"), STAT_InterestManagerComputationBox, STATGROUP_SpatialNet);
+DECLARE_CYCLE_STAT(TEXT("InterestManagerComputeBoxSSE"), STAT_InterestManagerComputationBoxSSE, STATGROUP_SpatialNet);
+DECLARE_CYCLE_STAT(TEXT("InterestManagerComputeSort"), STAT_InterestManagerComputationSort, STATGROUP_SpatialNet);
+
+namespace SpatialGDK
+{
+FTagComponentStorage::FTagComponentStorage(Worker_ComponentId InComponentId)
+	: TagComponentId(InComponentId)
+{
+	Components.Add(TagComponentId);
+}
+
+void FTagComponentStorage::OnComponentAdded(Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentData* Data)
+{
+	if (ensureAlways(TagComponentId == ComponentId))
+	{
+		TaggedObjects.Add(EntityId);
+		Modified.Add(EntityId);
+	}
+}
+
+void FTagComponentStorage::OnRemoved(Worker_EntityId EntityId)
+{
+	TaggedObjects.Remove(EntityId);
+	Modified.Remove(EntityId);
+}
+
+void FTagComponentStorage::OnUpdate(Worker_EntityId EntityId, Worker_ComponentId InComponentId, Schema_ComponentUpdate* Update)
+{
+	if (!ensure(false))
+	{
+		return;
+	}
+	Modified.Add(EntityId);
+}
+
+struct FInterestManager::FBroadphaseImpl
+{
+	FBroadphaseImpl() {}
+
+	void Add(uint32 Slot, FVector Position) {}
+
+	void Add(uint32 Slot, FBox2D Box) {}
+
+	void Update(uint32 Slot, FVector Position) {}
+
+	void SwapEntries(int32 Slot1, int32 Slot2) {}
+
+	void Remove(uint32 Slot) {}
+
+	void ComputeBoxVisibility(const TArray<FBox2D>& Regions, uint64* OutVisibility) {}
+};
+
+FInterestManager::~FInterestManager() = default;
+
+FInterestManager::FInterestManager(InterestFactory& InInterestF, FSpatialPositionStorage& InPositions,
+								   FAlwaysRelevantStorage& InAlwaysRelevant, FServerAlwaysRelevantStorage& InServerAlwaysRelevant)
+	: InterestF(InInterestF)
+	, Positions(InPositions)
+	, AlwaysRelevant(InAlwaysRelevant)
+	, ServerAlwaysRelevant(InServerAlwaysRelevant)
+{
+	SlotMap.Reserve(1 << 16);
```

Comment:
This is a SoA contruct for optimisation reasons I assume?

Reply:
SoA?

Reply:
[Structure of Arrays](https://en.wikipedia.org/wiki/AoS_and_SoA) as opposed to an Array of Structures.

Reply:
Yep 👍